### PR TITLE
Add in support for explode on maps

### LINKED ISSUE #3110: [FEA] Support Map as input to explode and pos_explode
**Is your feature request related to a problem? Please describe.**
This comes from an audit of missing functionality for nested types. Spark supports explode and pos_explode on maps. We should be able to support them too as we store a map as an array of key/value structs so we should be able to do the explode on the array in the map, and then pull out the key/value columns from the resulting struct. This should require no changes to cudf to support this.

### PR DESCRIPTION
**Pull request by revans2**
This fixes #3110

### CODE REVIEW

**Comment by revans2** [comment]
build

File: sql-plugin/src/main/java/com/nvidia/spark/rapids/GpuColumnVector.java [hunk 1]
```diff
@@ -695,7 +695,10 @@ static boolean typeConversionAllowed(ColumnView cv, DataType colType) {
 
   static boolean typeConversionAllowed(Table table, DataType[] colTypes, int startCol, int endCol) {
     final int numColumns = endCol - startCol;
-    assert numColumns == colTypes.length: "The number of columns and the number of types don't match";
+    assert numColumns == colTypes.length: "The number of columns and the number of types don't " +
+        "match. expected " + colTypes.length + " but found " + numColumns + ". (" + table +
```
Comment (jlowe) [inline sql-plugin/src/main/java/com/nvidia/spark/rapids/GpuColumnVector.java:699]:
```suggestion
        "match. Expected " + colTypes.length + " but found " + numColumns + ". (" + table +
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGenerateExec.scala [hunk 2]
```diff
@@ -228,24 +228,121 @@ abstract class GpuExplodeBase extends GpuUnevaluableUnaryExpression with GpuGene
...
+      fixChildValidity: Boolean): Table = {
+    val numPos = if (position) 1 else 0
+    // scalastyle:off line.size.limit
+    // The input will look like the following, and we just want to expand the key, value in the
+    // struct into separate columns
+    // INDEX [0, genOffset)| genOffset   | genOffset + numPos | [genOffset + numPos + 1, exploded.getNumberOfColumns)
+    // SOME INPUT COLUMNS  | POS COLUMN? | STRUCT(KEY, VALUE) | MORE INPUT COLUMNS
+    // scalastyle:on line.size.limit
+    val structPos = genOffset + numPos
+    withResource (ArrayBuffer.empty[ColumnVector] ) { newColumns =>
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGenerateExec.scala:311]: Nit: some odd extra whitespace here.
```suggestion
    withResource(ArrayBuffer.empty[ColumnVector]) { newColumns =>
```

**Review by jlowe — APPROVED** [review]
Minor nits, lgtm.

**Comment by revans2** [comment]
build

**Review by jlowe — APPROVED** [review]

**Pull request merged**